### PR TITLE
[WIP] Authorize listing of buckets (fixes #454)

### LINKED ISSUE #454: Authenticated user should be allowed to browse his buckets by default
Kinto config:

```
[app:main]
use = egg:kinto

# Required by integration tests
kinto.flush_endpoint_enabled = true

# Required by basic auth
kinto.userid_hmac_secret = a-secret-string

# Allow bbrowsing all buckets
# kinto.bucket_read_principals = system.Authenticated

[server:main]
use = egg:waitress#main
host = 0.0.0.0
port = 8888
```

Starting server

```
./.venv/bin/pserve test/kinto.ini
Starting server in PID 3230.
serving on http://0.0.0.0:8888
```

Listing buckets:

```
$ http :8888/v1/buckets -a a:
HTTP/1.1 403 Forbidden
Access-Control-Expose-Headers: Content-Length, Expires, Alert, Retry-After, Last-Modified, Total-Records, ETag, Pragma, Cache-Control, Backoff, Next-Page
Content-Length: 95
Content-Type: application/json; charset=UTF-8
Date: Tue, 23 Feb 2016 08:58:07 GMT
Server: waitress

{
    "code": 403, 
    "errno": 121, 
    "error": "Forbidden", 
    "message": "This user cannot access this resource."
}
```

This is quite surprising, as one might expect being able to at least list his own default bucket in there.

If you create a custom bucket, then you're suddenly allowed to browser buckets:

```
$ http PUT :8888/v1/buckets/foo -a a:
HTTP/1.1 201 Created
Access-Control-Expose-Headers: Retry-After, Content-Length, Alert, Backoff
Content-Length: 154
Content-Type: application/json; charset=UTF-8
Date: Tue, 23 Feb 2016 09:03:09 GMT
Etag: "1456218189534"
Last-Modified: Tue, 23 Feb 2016 09:03:09 GMT
Server: waitress

{
    "data": {
        "id": "foo", 
        "last_modified": 1456218189534
    }, 
    "permissions": {
        "write": [
            "basicauth:a2125d3ab6f6b9509e1bc87932010af71cebfd50424e972934a0e4b5a1047f42"
        ]
    }
}

$ http :8888/v1/buckets -a a:        
HTTP/1.1 200 OK
Access-Control-Expose-Headers: Content-Length, Expires, Alert, Retry-After, Last-Modified, Total-Records, ETag, Pragma, Cache-Control, Backoff, Next-Page
Cache-Control: no-cache
Content-Length: 53
Content-Type: application/json; charset=UTF-8
Date: Tue, 23 Feb 2016 09:03:17 GMT
Etag: "1456218189534"
Last-Modified: Tue, 23 Feb 2016 09:03:09 GMT
Server: waitress
Total-Records: 1

{
    "data": [
        {
            "id": "foo", 
            "last_modified": 1456218189534
        }
    ]
}
```

TL;DR: I think users should be able to list buckets they have access to by default.

### PR DESCRIPTION
**Pull request by leplatrem**
* [ ] Validate approach taken
* [ ] Wait for https://github.com/mozilla-services/cliquet/pull/693
* [ ] Decide whether it could be generalized (to collections, groups etc.)

@almet r?

### CODE REVIEW

File: kinto/authorization.py [hunk 1]
```diff
@@ -143,6 +143,15 @@ class AuthorizationPolicy(cliquet_authorization.AuthorizationPolicy):
     def get_bound_permissions(self, *args, **kwargs):
         return build_permissions_set(*args, **kwargs)
 
+    def permits_shared_resource(self, context, principals, permission):
```
Comment (almet) [inline kinto/authorization.py]: I don't understand where this method is called?

**Comment by leplatrem** [comment]
The issue #454 was transformed into a pull-request. Closing this.
